src/CabalHelper/Runtime/Main.hs

```diff
@@ -554,14 +611,51 @@ removeInplaceDeps _v lbi pd clbi = let
 
 
 #if CH_MIN_VERSION_Cabal(2,0,0)
-recursiveIncludeDirs :: Map.Map UnitId ([UnitId],[FilePath], [(OpenUnitId, ModuleRenaming)])
-                     -> UnitId -> ([FilePath],[(OpenUnitId, ModuleRenaming)])
-recursiveIncludeDirs includeDirs unit = go ([],[]) [unit]
+recursiveDepInfo lbi v distdir = do
+  includeDirs <- componentsMap lbi v distdir $ \c clbi bi -> do
+    return (componentUnitId clbi
+           , ( componentInternalDeps clbi
+             , hsSourceDirs bi
+             , componentIncludes clbi
+             , componentEntrypoints c))
+  return $ Map.fromList $ map snd includeDirs
+
+type SubDeps = ([UnitId], [FilePath], [(OpenUnitId, ModuleRenaming)], ChEntrypoint)
+
+recursiveIncludeDirs :: Map.Map UnitId SubDeps
+                     -> UnitId -> ([FilePath], [(OpenUnitId, ModuleRenaming)]
+                                  , ChEntrypoint)
+recursiveIncludeDirs includeDirs unit = go ([],[],Nothing) [unit]
+  where
+    go (afp,aci,amep) [] = (afp,aci,maybe (error "recursiveIncludeDirs:no ChEntrypoint") id amep)
+    go acc@(afp,aci,amep) (u:us) = case Map.lookup u includeDirs of
+      Nothing -> go acc us
+      Just (us',sfp,sci,sep) -> go (afp++sfp,aci++sci,combineEp amep sep) (us++us')
+
+needsBuildOutput :: Map.Map UnitId SubDeps
+                     -> UnitId -> NeedsBuildOutput
```

Comment:
This isn't aligned, looks like it would fit in a single 80char line anyways. If so remove the line break.